arrows/vxl/hashed_image_classifier.cxx

```diff
@@ -0,0 +1,498 @@
+template < typename FeatureType, typename OutputType >
+void
+hashed_image_classifier< FeatureType, OutputType >
+::generate_weight_image( input_image_t const& src,
+                         weight_image_t& dst,
+                         unsigned const& feature_id ) const
+{
+  dst.set_size( src.ni(), src.nj() );
+  dst.fill( 0 );
+
+  weight_t** const feature_weights{ &model_->feature_weights[ 0 ] };
+
+  for( unsigned j = 0; j < src.nj(); ++j )
+  {
+    for( unsigned i = 0; i < src.ni(); ++i )
+    {
+      dst( i, j ) = feature_weights[ feature_id ][ src( i, j ) ];
+    }
+  }
+}
+
+// ----------------------------------------------------------------------------
+template < typename FeatureType, typename OutputType >
+void
+hashed_image_classifier< FeatureType, OutputType >
+::set_model( model_sptr_t external_model )
+{
+  if( !external_model->is_valid() )
+  {
+    LOG_ERROR( logger, "Input model invalid" );
+  }
+
+  model_ = external_model;
+}
+
+// ----------------------------------------------------------------------------
+template < typename FloatType >
+hashed_image_classifier_model< FloatType >
+
+::hashed_image_classifier_model( self_t const& other )
+{
+  *this = other;
+}
+
+// ----------------------------------------------------------------------------
+template < typename FloatType >
+bool
+hashed_image_classifier_model< FloatType >
+::is_valid() const
+{
+  if( num_features > 0 &&
+      num_features == max_feature_value.size() &&
+      num_features == feature_weights.size() &&
+      weights.size() > num_features )
+  {
+    return true;
+  }
+
+  LOG_ERROR( logger, "num_features: " << num_features );
+
+  return false;
+}
+
+// ----------------------------------------------------------------------------
+template < typename FloatType >
+void
+hashed_image_classifier_model< FloatType >
+::reset( unsigned feature_count, unsigned entries_per_feature )
+{
+  this->num_features = feature_count;
+
+  this->max_feature_value =
+    std::vector< unsigned >( feature_count, entries_per_feature );
+
+  this->weights =
+    std::vector< weight_t >( feature_count * entries_per_feature, 0.0 );
+
+  this->feature_weights =
+    std::vector< weight_t* >( feature_count );
+
+  for( unsigned i = 0; i < feature_count; ++i )
+  {
+    this->feature_weights[ i ] = &( this->weights[ 0 ] ) +
+                                 i * entries_per_feature;
+  }
+}
+
+// ----------------------------------------------------------------------------
+template < typename FloatType >
+void
+hashed_image_classifier_model< FloatType >
+::normalize( weight_t total_weight )
+{
+  double weight_magnitude_sum{ 0 };
+
+  for( auto const i : vital::range::iota( this->weights.size() ) )
+  {
+    weight_magnitude_sum += std::fabs( this->weights[ i ] );
+  }
+
+  if( weight_magnitude_sum != 0 )
+  {
+    weight_t norm_factor =
+      static_cast< weight_t >( 1.0 / ( weight_magnitude_sum * total_weight ) );
```

Comment:
Oops, over length :slightly_smiling_face:. `weight_magnitude_sum` might be overkill, `sum` is probably fine and would make this fit.